### PR TITLE
Miscellaneous Performance Improvements Copied from Azure SDK for Java

### PR DESCRIPTION
**Pull request by alzimmermsft**
This PR contains miscellaneous performance improvements that were added to Azure SDK for Java that are application to Azure SDK for Android.

### CODE REVIEW

File: sdk/core/azure-core-http/src/main/java/com/azure/android/core/http/util/UrlBuilder.java [hunk 1]
```diff
@@ -14,6 +14,7 @@
  */
 public final class UrlBuilder {
     private static final Map<String, UrlBuilder> PARSED_URLS = new HashMap<>();
+    private static final int MAX_CACHE_SIZE = 10000;
```
Comment (alzimmermsft) [inline sdk/core/azure-core-http/src/main/java/com/azure/android/core/http/util/UrlBuilder.java:17]: @anuchandy should we lower the maximum cache size given the more restrictive environment Android generally runs on

Comment (vcolin7) [reply sdk/core/azure-core-http/src/main/java/com/azure/android/core/http/util/UrlBuilder.java:17]: I don't know how often users would make use of this cache for new URLs. Do we have numbers on what normally happens in the Java SDK?

Comment (alzimmermsft) [reply sdk/core/azure-core-http/src/main/java/com/azure/android/core/http/util/UrlBuilder.java:17]: At least in the Java SDKs this is used in RestProxy to generated the injected request URL. For many services that use the same base URL this doesn't grow often, but services like Storage where each blob has a unique base URL this is used heavily and has resulted in OOMs in the past.

Comment (anuchandy) [reply sdk/core/azure-core-http/src/main/java/com/azure/android/core/http/util/UrlBuilder.java:17]: Sizing cache is not easy; here, as Alan pointed out, it depends on how likely specific services (that RestProxy proxies) tend to reuse the URLs. I'm fine to start with this number.

**Review by anuchandy — APPROVED** [review]
Thanks, Alan, LGTM

File: sdk/core/azure-core/src/main/java/com/azure/android/core/util/DateTimeRfc1123.java [hunk 2]
```diff
@@ -48,6 +55,170 @@ public OffsetDateTime getDateTime() {
...
+     * @return The specified substring.
+     *
+     * @throws DateTimeException If the processing character is not digit character.
+     */
+    private static int parseInt(final CharSequence date, final int beginIndex, final int endIndex) {
+        int num = 0;
+        for (int i = beginIndex; i < endIndex; i++) {
+            final char c = date.charAt(i);
+            if (c < '0' || c > '9') {
+                throw new DateTimeException("Invalid date time: " + date);
```
Comment (vcolin7) [inline sdk/core/azure-core/src/main/java/com/azure/android/core/util/DateTimeRfc1123.java:100]: We should probably [follow Java](https://github.com/Azure/azure-sdk-for-java/blob/042aa47cf0f30dec6c06d9534d9840fb197a4d06/sdk/core/azure-core/src/main/java/com/azure/core/util/DateTimeRfc1123.java#L110) here and log these exceptions.
```suggestion
                throw LOGGER.logExceptionAsError(new DateTimeException("Invalid date time: " + date));
```

File: sdk/core/azure-core/src/main/java/com/azure/android/core/util/DateTimeRfc1123.java [hunk 3]
```diff
@@ -48,6 +55,170 @@ public OffsetDateTime getDateTime() {
...
+        switch (date.charAt(beginIndex)) {
+            case 'J':
+                // Jan, Jun, Jul
+                switch (date.charAt(beginIndex + 1)) {
+                    case 'a': return 1; // Jan
+                    case 'u':
+                        switch (date.charAt(beginIndex + 2)) {
+                            case 'n': return 6; // Jun
+                            case 'l': return 7; // Jul
+                            default: throw new IllegalArgumentException("Unknown month " + date);
```
Comment (vcolin7) [inline sdk/core/azure-core/src/main/java/com/azure/android/core/util/DateTimeRfc1123.java:130]:
```suggestion
                            default: throw LOGGER.logExceptionAsError(
                                new IllegalArgumentException("Unknown month " + date));
```

File: sdk/core/azure-core/src/main/java/com/azure/android/core/util/DateTimeRfc1123.java [hunk 4]
```diff
@@ -48,6 +55,170 @@ public OffsetDateTime getDateTime() {
...
+                        }
+                    default: throw new IllegalArgumentException("Unknown month " + date);
```
Comment (vcolin7) [inline sdk/core/azure-core/src/main/java/com/azure/android/core/util/DateTimeRfc1123.java:132]:
```suggestion
                    default: throw LOGGER.logExceptionAsError(new IllegalArgumentException("Unknown month " + date));
```

File: sdk/core/azure-core/src/main/java/com/azure/android/core/util/DateTimeRfc1123.java [hunk 5]
```diff
@@ -48,6 +55,170 @@ public OffsetDateTime getDateTime() {
...
+                }
+            case 'F': return 2; // Feb
+            case 'M':
+                // Mar, May
+                switch (date.charAt(beginIndex + 2)) {
+                    case 'r': return 3; // Mar
+                    case 'y': return 5; // May
+                    default: throw new IllegalArgumentException("Unknown month " + date);
```
Comment (vcolin7) [inline sdk/core/azure-core/src/main/java/com/azure/android/core/util/DateTimeRfc1123.java:140]:
```suggestion
                    default: throw LOGGER.logExceptionAsError(new IllegalArgumentException("Unknown month " + date));
```

File: sdk/core/azure-core/src/main/java/com/azure/android/core/util/DateTimeRfc1123.java [hunk 6]
```diff
@@ -48,6 +55,170 @@ public OffsetDateTime getDateTime() {
...
+                }
+            case 'A':
+                // Apr, Aug
+                switch (date.charAt(beginIndex + 2)) {
+                    case 'r': return 4; // Apr
+                    case 'g': return 8; // Aug
+                    default: throw new IllegalArgumentException("Unknown month " + date);
```
Comment (vcolin7) [inline sdk/core/azure-core/src/main/java/com/azure/android/core/util/DateTimeRfc1123.java:147]:
```suggestion
                    default: throw LOGGER.logExceptionAsError(new IllegalArgumentException("Unknown month " + date));
```

File: sdk/core/azure-core/src/main/java/com/azure/android/core/util/DateTimeRfc1123.java [hunk 7]
```diff
@@ -48,6 +55,170 @@ public OffsetDateTime getDateTime() {
...
+                }
+            case 'S': return 9; //Sep
+            case 'O': return 10; // Oct
+            case 'N': return 11; // Nov
+            case 'D': return 12; // Dec
+            default: throw new IllegalArgumentException("Unknown month " + date);
```
Comment (vcolin7) [inline sdk/core/azure-core/src/main/java/com/azure/android/core/util/DateTimeRfc1123.java:153]:
```suggestion
            default: throw LOGGER.logExceptionAsError(new IllegalArgumentException("Unknown month " + date));
```

File: sdk/core/azure-core/src/main/java/com/azure/android/core/util/DateTimeRfc1123.java [hunk 8]
```diff
@@ -14,6 +19,8 @@
  * during serialization and deserialization.
  */
 public final class DateTimeRfc1123 {
+    private static final ClientLogger LOGGER = new ClientLogger(DateTimeRfc1123.class);
```
Comment (vcolin7) [inline sdk/core/azure-core/src/main/java/com/azure/android/core/util/DateTimeRfc1123.java:22]: I recall an old conversation with @anuchandy about avoiding using static loggers in Android but I can't really recall the actual reason. I'll ask him separately and come back to check if we'd prefer to go with something else.

File: sdk/core/azure-core-jackson/src/main/java/com/azure/android/core/serde/jackson/JacksonSerder.java [hunk 9]
```diff
@@ -40,7 +40,11 @@
  */
 public final class JacksonSerder {
     private static final Pattern PATTERN = Pattern.compile("^\"*|\"*$");
-    private static JacksonSerder jacksonSerder;
+
+    private static final int CACHE_SIZE_LIMIT = 10000;
```
Comment (vcolin7) [inline sdk/core/azure-core-jackson/src/main/java/com/azure/android/core/serde/jackson/JacksonSerder.java:44]: Based on @anuchandy's answer [here](https://github.com/Azure/azure-sdk-for-android/pull/972/files#r868513723), I think this value looks good but we should also pay close attention to it.

**Pull request merged**